pkg/storage/google_cloud_storage.go

```diff
@@ -0,0 +1,218 @@
+	client *storage.Client
+	cfg    *Config
+}
+
+// NewGoogleCloudStorage creates a new GoogleCloudStorage client.
+func NewGoogleCloudStorage(ctx context.Context, opts ...Option) (*GoogleCloudStorage, error) {
+	cfg := &Config{
+		initialRetryDelay: 1 * time.Second,
+		maxRetryDelay:     20 * time.Second,
+		retryMultiplier:   2,
+		retryTimeout:      60 * time.Second,
+	}
+
+	for _, opt := range opts {
+		if opt != nil {
+			cfg = opt(cfg)
+		}
+	}
+
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create google cloud storage client: %w", err)
+	}
+
+	return &GoogleCloudStorage{
+		cfg:    cfg,
+		client: client,
+	}, nil
+}
+
+// objectHandleWithRetries retrieves an object handle configured with a retry mechanism.
+func (s *GoogleCloudStorage) objectHandleWithRetries(ctx context.Context, bucket, name string) (*storage.ObjectHandle, context.Context, context.CancelFunc) {
+	o := s.client.Bucket(bucket).Object(name).Retryer(
+		storage.WithBackoff(gax.Backoff{
+			Initial:    s.cfg.initialRetryDelay,
+			Max:        s.cfg.maxRetryDelay,
+			Multiplier: s.cfg.retryMultiplier,
+		}),
+		storage.WithPolicy(storage.RetryAlways),
+	)
+	ctx, cancel := context.WithTimeout(ctx, s.cfg.retryTimeout)
+	return o, ctx, cancel
+}
+
+// makeUploadConfig creates default upload config and overwrites with the user
+// provided upload options.
+func makeUploadConfig(contentLength int, opts []UploadOption) *uploadConfig {
+	defaultChunkSize := 16 * MiB // default for cloud storage client
+
+	// we can send smaller files in one request
+	if contentLength <= 16*MiB {
+		defaultChunkSize = contentLength + 256
+	}
+
+	cfg := &uploadConfig{
+		chunkSize:          defaultChunkSize,
+		cacheMaxAgeSeconds: 86400, // 1 day
+	}
+
+	for _, opt := range opts {
+		if opt != nil {
+			cfg = opt(cfg)
+		}
+	}
+
+	cacheControl := fmt.Sprintf("public, max-age=%d", cfg.cacheMaxAgeSeconds)
+	if cfg.cacheMaxAgeSeconds == 0 {
+		cacheControl = "no-cache, max-age=0"
+	}
+	cfg.cacheControl = cacheControl
+
+	return cfg
+}
+
+func (s *GoogleCloudStorage) UploadObject(ctx context.Context, bucket, name string, contents []byte, opts ...UploadOption) error {
+	cfg := makeUploadConfig(len(contents), opts)
+
+	o, ctx, cancel := s.objectHandleWithRetries(ctx, bucket, name)
+	defer cancel()
+
+	writer := o.If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	defer writer.Close()
+
+	writer.CacheControl = cfg.cacheControl
+	writer.ChunkSize = cfg.chunkSize
+
+	if cfg.contentType != "" {
+		writer.ContentType = cfg.contentType
+	}
+
+	if cfg.metadata != nil {
+		m := make(map[string]string, len(cfg.metadata))
+
+		for k, v := range cfg.metadata {
+			m[k] = v
+		}
+
+		writer.Metadata = m
+	}
+
+	if _, err := writer.Write(contents); err != nil {
+		return fmt.Errorf("failed to write data: %w", err)
+	}
+
+	return nil
+}
+
+func (s *GoogleCloudStorage) DownloadObject(ctx context.Context, bucket, name string) (io.ReadCloser, error) {
```

Comment:
Doc: it's up to the caller to close the reader

Reply:
Should this go in the interface doc or the implementation....or both?